# Лабораторная работа №6

In [33]:
%load_ext sql
%sql postgresql://AAK:@127.0.0.1/students
%config SqlMagic.displaylimit = 20

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


1.  ## Задание 1
    > Проведите следующий эксперимент:
    > Создайте простую таблицу, содержащую три поля – суррогатный ключ и два атрибута, содержащие строки.

In [23]:
%%sql
CREATE TABLE Test(
	id SERIAL PRIMARY KEY,
	CODE_1 VARCHAR(64),
	CODE_2 VARCHAR(64)
);

Running query in 'postgresql://AAK:***@127.0.0.1/students'

++
||
++
++

> С помощью следующего скрипта заполните таблицу данными.

In [24]:
%%sql
DO
$$
BEGIN
FOR i IN 1..1000000 LOOP
    INSERT INTO Test(code_1,code_2) VALUES(md5(random()::text), md5(random()::text));
END LOOP;
END
$$ language plpgsql;

Running query in 'postgresql://AAK:***@127.0.0.1/students'

++
||
++
++

> С помощью функций EXPLAIN ANALYZE в PostgreSQL возможно производить анализ запросов и вычислять затрачиваемое время на его выполнение. Например,

In [25]:
%%sql
EXPLAIN ANALYZE SELECT * FROM student;

Running query in 'postgresql://AAK:***@127.0.0.1/students'

3 rows affected.

+------------------------------------------------------------------------------------------------------+
|                                              QUERY PLAN                                              |
+------------------------------------------------------------------------------------------------------+
| Seq Scan on student  (cost=0.00..4.63 rows=163 width=90) (actual time=0.046..0.095 rows=189 loops=1) |
|                                       Planning Time: 0.086 ms                                        |
|                                       Execution Time: 0.119 ms                                       |
+------------------------------------------------------------------------------------------------------+

> Добавьте в таблицу Test одно значение, измерив время данной операции.
> Далее измерьте время выполнения запроса,
> выводящего содержимого таблицы в отсортированном виде по столбцу CODE_1.
>
> Добавьте индекс на столбец CODE_1. Повторите предыдущие две операции.
> Сравните полученное время. Во сколько раз оно изменилось? Результаты вычисления занесите в таблицу.

Добавим одно значение в таблицу:

In [26]:
%%sql
EXPLAIN ANALYZE INSERT INTO Test(code_1,code_2) VALUES(1, 1);

Running query in 'postgresql://AAK:***@127.0.0.1/students'

4 rows affected.

+--------------------------------------------------------------------------------------------+
|                                         QUERY PLAN                                         |
+--------------------------------------------------------------------------------------------+
| Insert on test  (cost=0.00..0.01 rows=0 width=0) (actual time=0.049..0.050 rows=0 loops=1) |
|   ->  Result  (cost=0.00..0.01 rows=1 width=296) (actual time=0.014..0.014 rows=1 loops=1) |
|                                  Planning Time: 0.054 ms                                   |
|                                  Execution Time: 0.087 ms                                  |
+--------------------------------------------------------------------------------------------+

Измерим время выполнения запроса:

In [27]:
%%sql
EXPLAIN ANALYZE SELECT * FROM Test ORDER BY CODE_1;

Running query in 'postgresql://AAK:***@127.0.0.1/students'

11 rows affected.

+------------------------------------------------------------------------------------------------------------------------------------+
|                                                             QUERY PLAN                                                             |
+------------------------------------------------------------------------------------------------------------------------------------+
|        Gather Merge  (cost=43129.77..73139.47 rows=257208 width=296) (actual time=2987.198..5382.966 rows=1000001 loops=1)         |
|                                                          Workers Planned: 2                                                        |
|                                                         Workers Launched: 2                                                        |
|            ->  Sort  (cost=42129.75..42451.26 rows=128604 width=296) (actual time=2893.862..3821.842 rows=333334 loops=3)          |
|                                                              Sort Key: code_1                                                      |
|                                                 Sort Method: external merge  Disk: 25720kB                                         |
|                                           Worker 0:  Sort Method: external merge  Disk: 26856kB                                    |
|                                           Worker 1:  Sort Method: external merge  Disk: 25704kB                                    |
|         ->  Parallel Seq Scan on test  (cost=0.00..13632.04 rows=128604 width=296) (actual time=0.037..74.216 rows=333334 loops=3) |
|                                                      Planning Time: 0.113 ms                                                       |
|                                                    Execution Time: 5456.078 ms                                                     |
+------------------------------------------------------------------------------------------------------------------------------------+

Добавим индекс:

In [28]:
%%sql
CREATE INDEX code_1_idx ON Test (code_1);

Running query in 'postgresql://AAK:***@127.0.0.1/students'

++
||
++
++

Повторим операции:

In [29]:
%%sql
EXPLAIN ANALYZE INSERT INTO Test(code_1,code_2) VALUES(1, 1);

Running query in 'postgresql://AAK:***@127.0.0.1/students'

4 rows affected.

+--------------------------------------------------------------------------------------------+
|                                         QUERY PLAN                                         |
+--------------------------------------------------------------------------------------------+
| Insert on test  (cost=0.00..0.01 rows=0 width=0) (actual time=2.027..2.028 rows=0 loops=1) |
|   ->  Result  (cost=0.00..0.01 rows=1 width=296) (actual time=0.011..0.011 rows=1 loops=1) |
|                                  Planning Time: 0.049 ms                                   |
|                                  Execution Time: 2.067 ms                                  |
+--------------------------------------------------------------------------------------------+

*Execution time* увеличился

In [30]:
%%sql
EXPLAIN ANALYZE SELECT * FROM Test ORDER BY CODE_1;

Running query in 'postgresql://AAK:***@127.0.0.1/students'

3 rows affected.

+-------------------------------------------------------------------------------------------------------------------------------------+
|                                                              QUERY PLAN                                                             |
+-------------------------------------------------------------------------------------------------------------------------------------+
| Index Scan using code_1_idx on test  (cost=0.42..93224.44 rows=1000001 width=296) (actual time=0.087..977.950 rows=1000002 loops=1) |
|                                                       Planning Time: 0.124 ms                                                       |
|                                                     Execution Time: 1035.395 ms                                                     |
+-------------------------------------------------------------------------------------------------------------------------------------+

*Execution time* значительно сократился

**Вывод**: добавлять записи дольше, а искать быстрее.

|                    | insert | select |
| ------------------ | ------ | ------ |
| До индекса (мс)    | 0.087  | 5456   |
| После индекса (мс) | 2.067  | 1035   |
| Прирост (раз)      | 0.042  | 5.27   |

2.  ## Задание 2.

> Составьте запрос к таблице Test, выводящий все строки в отсортированном порядке,
> в которых столбец CODE_1 начинается с символа ‘a’. Проанализируйте полученный запрос и объясните результат.

In [31]:
%%sql
EXPLAIN ANALYZE SELECT * FROM Test
  WHERE CODE_1 ~ '^a'
  ORDER BY CODE_1

Running query in 'postgresql://AAK:***@127.0.0.1/students'

13 rows affected.

+----------------------------------------------------------------------------------------------------------------------------------+
|                                                            QUERY PLAN                                                            |
+----------------------------------------------------------------------------------------------------------------------------------+
|          Gather Merge  (cost=20400.81..26293.35 rows=50504 width=70) (actual time=489.894..594.561 rows=62665 loops=1)           |
|                                                         Workers Planned: 2                                                       |
|                                                        Workers Launched: 2                                                       |
|             ->  Sort  (cost=19400.78..19463.91 rows=25252 width=70) (actual time=386.539..390.144 rows=20888 loops=3)            |
|                                                             Sort Key: code_1                                                     |
|                                                  Sort Method: quicksort  Memory: 3667kB                                          |
|                                            Worker 0:  Sort Method: quicksort  Memory: 2617kB                                     |
|                                            Worker 1:  Sort Method: quicksort  Memory: 2740kB                                     |
|         ->  Parallel Seq Scan on test  (cost=0.00..17554.34 rows=25252 width=70) (actual time=0.053..220.908 rows=20888 loops=3) |
|                                                     Filter: ((code_1)::text ~ '^a'::text)                                        |
|                                                         Rows Removed by Filter: 312446                                           |
|                                                     Planning Time: 1.891 ms                                                      |
|                                                    Execution Time: 599.521 ms                                                    |
+----------------------------------------------------------------------------------------------------------------------------------+

In [34]:
%%sql
SELECT * FROM Test
  WHERE CODE_1 ~ '^a'
  ORDER BY CODE_1

Running query in 'postgresql://AAK:***@127.0.0.1/students'

62665 rows affected.

+--------+----------------------------------+----------------------------------+
|   id   |              code_1              |              code_2              |
+--------+----------------------------------+----------------------------------+
| 147695 | a0000705fda10a89f7200f6096108ff8 | 44246d744ce5aadac4f8178d8c8cbc51 |
| 258578 | a000089c5dd3463b525c65153db6cbc4 | 86e55d990e58335b0076e5d5f9764802 |
| 121872 | a0000e0ef9d728cd410d1ef7e1e45336 | 2276aa40398d47f2d687615df36578aa |
| 992198 | a000237d643c732ba82ebfa2ba90dafc | cb7b06b57177f0d90e646d8056083b7d |
| 167720 | a0005c2a4741d1ab29a129bc1cfe1a2a | b9ef748c415b966215c4d73dcd43b7d0 |
| 907780 | a000a07a4626f5f7dceb09fce88ffdbd | f1e7f0d87bf14e03a624eda0a830235f |
| 302215 | a000b1769bffe82fdff12119ef06a130 | 72ad0115197d5e85da48322ae8bc2094 |
| 871061 | a000badb9cae5d0508d3b1a6b8639fee | 3c979e79e97a82383e6ed8bad945705d |
| 265089 | a000bdc9ba2666e9a127a3911187835d | f9d8409d64faf9b302b374581a978ee7 |
| 225925 | a000c5810b4adb334165caaa4ba841ac | c9529895f751655bfc512a1d83042409 |
| 30870  | a000e4fb03f02ea23f25db826344017a | 96607412af10d8274eb99350f758c6ba |
| 862917 | a000f0e4ea40246f828d9fac7e3ab5a4 | 0a9943c1a3efa894c43f32f77229a74b |
| 168019 | a000f4dd018ce51de67bd2e076f93cff | 73def6bd7024ba8814c694766fd8b918 |
| 296952 | a000f927655ea4bfe6b3bf175960af94 | 9eb5a02ae28a863d1b79f7f0b164adb6 |
| 850792 | a000f9f80eb6205fc5af90af0016f3bf | b47c1330ccbeeef1360327dee576996d |
| 421274 | a00128ba34cd0307c0cbbb4602fb7946 | f396f53bdd5543c629c7af3c03ba9013 |
| 677780 | a0013629bd4720953f579d1336b36d10 | c759c101b2b22a3e8ed116c347246426 |
| 601314 | a001439af40e4e67b41413d4c2fdfa70 | 3c286d7aae4d7b412bbafced95e34d45 |
| 336928 | a00151cb296f34cf0f2067f3e0787039 | 3c174c99e9a4c4fc47fc5ac8b27b9280 |
| 495269 | a00160d2f9fa8ae1113fb7ac98afa7bf | 9ae2c7a0788d56e61acb719b2e2875e5 |
+--------+----------------------------------+----------------------------------+
Truncated to displaylimit of 20.

3.  ## Задание 3.

    > Проанализируйте учебную базу данных и проиндексируйте одно из полей любой таблицы. Объясните свой выбор.

Тк ключи в индексах сгруппированы по страницам, то нужно создать индекс для колонки,
в которой много значений. Иначе из-за большого размера страницы может получиться, что дерево
содержит только один лист. Много студентов, оценок, преподавателей.
Желательно, чтобы индекс менялся нечасто, но часто просматривался, поэтому оценки не подойдут.
У студентов и преподавателей есть *id* и по нему часто проводится поиск, поэтому индекс по *id*
может ускорить поиск, но при запросах к этим полям, в основном, имеет смысл операция проверки на
равенство. То есть не используются полные возможности свойства сортированности ключей.
Хорошей целью для создания индекса кажется поле *birthday* в таблице *student*, тк часто будут
применять операции сравнения <, > к этому полю.

In [35]:
%%sql
CREATE INDEX birthday_idx ON student (birthday);

Running query in 'postgresql://AAK:***@127.0.0.1/students'

++
||
++
++

4.  ## Задание 4.

    > Предположим, преподаватель зашел в электронный журнал,
    > чтобы проставить студенту оценку за дисциплину.
    > Одновременно студент решил проверить, выставлена ли ему оценка.
    > Смоделируйте данную ситуацию, заключив действия студента и преподавателя в транзакции с
    > уровнем изоляции, в соответствии с вариантом.
    > Приведите пример аномалий, которые могут возникнуть.

    > Read committed

    *Read committed* - режим по умолчанию для транзакций в *Postgres*.

    Моделирование случая успешной и неуспешной записей в журнал смотри в [видео](/lab6/2023-11-09_23-26-04.mkv)

    В общем (но не в *postgres*) возможна аномалия *«Грязное» чтение*. То есть во время транзакции
    записи преподавателя возникла ошибка, и запись не случилась, а студент увидит у себя
    неправильную оценку.

    Или, например, аномалия *неповторяющегося чтения*. Если в одной транзакции студента мы получаем
    и оценку по предмету, и высчитываем средний балл, и между этими операциями возникла операция
    записи преподавателя, то средний балл не будет вычисляться из видимой оценки.

In [37]:
# %%sql
# BEGIN;
#   SELECT mark
#     FROM field_comprehension
#     WHERE student_id = 838389 AND field = 'fa706604-353c-3fc2-bc9c-53559dc4f79d';

#   UPDATE field_comprehension
#     SET mark = 5
#     WHERE student_id = 838389 AND field = 'fa706604-353c-3fc2-bc9c-53559dc4f79d';

#   SELECT mark
#     FROM field_comprehension
#     WHERE student_id = 838389 AND field = 'fa706604-353c-3fc2-bc9c-53559dc4f79d';

#     ROLLBACK
